crates/ruff_python_formatter/src/other/f_string_element.rs

```diff
@@ -0,0 +1,264 @@
+            let (dangling_open_parentheses_comments, trailing_format_spec_comments) =
+                dangling_item_comments.split_at(
+                    dangling_item_comments
+                        .partition_point(|comment| comment.start() < expression.start()),
+                );
+
+            let item = format_with(|f| {
+                let line_break_or_space = match expression.as_ref() {
+                    // If an expression starts with a `{`, we need to add a space before the
+                    // curly brace to avoid turning it into a literal curly with `{{`.
+                    //
+                    // For example,
+                    // ```python
+                    // f"{ {'x': 1, 'y': 2} }"
+                    // #  ^                ^
+                    // ```
+                    //
+                    // We need to preserve the space highlighted by `^`.
+                    Expr::Dict(_) | Expr::DictComp(_) | Expr::Set(_) | Expr::SetComp(_) => {
+                        Some(soft_line_break_or_space())
+                    }
+                    _ => None,
+                };
+
+                let f = &mut WithExprLocation::new(
+                    ExpressionLocation::InsideFString(self.context.quotes()),
+                    f,
+                );
+
+                line_break_or_space.fmt(f)?;
+
+                // If we're going to remove the soft line breaks, then there's a chance
+                // that there will be trailing commas in the formatted expression.
+                // Currently, it's difficult to remove that conditionally, because (TODO)
+                // So, we'll manually format the expression with the maximum line width
+                // and disabling the magic trailing comma. This is expensive so we've
+                // implemented some heuristics to avoid this in some cases.
+                if self.context.should_remove_soft_line_breaks()
+                    && !matches!(
+                        &**expression,
+                        Expr::BooleanLiteral(_)
+                            | Expr::BytesLiteral(_)
+                            | Expr::EllipsisLiteral(_)
+                            | Expr::IpyEscapeCommand(_)
+                            | Expr::Name(_)
+                            | Expr::NoneLiteral(_)
+                            | Expr::NumberLiteral(_)
+                            | Expr::StringLiteral(_)
+                    )
+                {
+                    let options = f
+                        .options()
+                        .clone()
+                        .with_line_width(NonZeroU16::MAX.into())
+                        .with_magic_trailing_comma(MagicTrailingComma::Ignore);
+                    let context =
+                        PyFormatContext::new(options, f.context().source(), comments.clone())
+                            .in_f_string(self.context.quotes());
```

Comment:
I think we have to reuse the outer context here or you lose information (are we in a prenthesized context? what's the indent level and potentially other new state that will be added). 

You can reuse the context by building the `VecBuffer` manually

```python
let formatted = {
	let mut buffer = VecBuffer::new(f.state_mut());
	write!(buffer, [expression.format()])?;
	buffer.into_vec()
};

let print_options = f.context.options().as_print_options();
let expression_str = Printer::new(f.context().source_code(), print_options).print(&formatted)?;
```

Reply:
We need to have different options (max line width, no magic trailing comma) which is why I choose to create a new context but I see your point. I'm using the following:

```rs
                    let context = f.context().clone().in_f_string(self.context.quotes());
                    let formatted = crate::format!(context, [expression.format()])?;
                    let printer_options = f
                        .options()
                        .clone()
                        .with_line_width(NonZeroU16::MAX.into())
                        .with_magic_trailing_comma(MagicTrailingComma::Ignore)
                        .as_print_options();
                    text(
                        Printer::new(f.context().source_code(), printer_options)
                            .print(formatted.document())?
                            .as_code(),
                    )
                    .fmt(f)?;
```

Reply:
This doesn't work because the options need to be updated for the current context as well.